### PR TITLE
[core] Introduce RollbackToTimestamp Procedure and Action

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose

Linked issue: close #xxx
Currently Rollback cannot rollback according timestamp which can make user more easy to rollback, as https://github.com/apache/paimon/pull/4333 suggest introduce RollbackToTimestamp to support for procedure and action.

### CODE REVIEW

File: docs/content/flink/procedures.md [hunk 1]
```diff
@@ -356,6 +356,28 @@ All available procedures are listed below.
...
+            <li>identifier: the target table identifier. Cannot be empty.</li>
+            <li>timestamp (Long): Roll back to the snapshot which earlier or equal than timestamp.</li>
+      </td>
+      <td>
+         -- for Flink 1.18<br/>
+         CALL sys.rollback_to_timestamp('default.T', 10)
+         -- for Flink 1.19 and later<br/>
+         CALL sys.rollback_to(`table` => 'default.T', snapshot_id => 1730292023000)
+      </td>
+   </tr>
```
Comment (LinMingQiang) [inline docs/content/flink/procedures.md:380]: rollback_to -> rollback_to_timestamp

snapshot_id -> timestamp

File: paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.paimon.table.Table;
+import org.apache.paimon.utils.Preconditions;
+
+import org.apache.flink.table.procedure.ProcedureContext;
+
+/**
+ * Rollback to timestamp procedure. Usage:
+ *
+ * <pre><code>
+ *  -- rollback to a snapshot
```
Comment (LinMingQiang) [inline paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java:34]: ditto

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/RollbackToTimestampProcedure.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                table -> {
+                    FileStoreTable fileStoreTable = (FileStoreTable) table;
+                    Snapshot snapshot =
+                            fileStoreTable.snapshotManager().earlierOrEqualTimeMills(timestamp);
+                    Preconditions.checkNotNull(
+                            snapshot,
+                            String.format("count not find snapshot earlier than %s", timestamp));
+                    fileStoreTable.rollbackTo(snapshot.id());
+                    InternalRow outputRow = newInternalRow(true);
+                    return new InternalRow[] {outputRow};
```
Comment (LinMingQiang) [inline paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/RollbackToTimestampProcedure.java:82]: ditto

File: paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    public String[] call(ProcedureContext procedureContext, String tableId, long timestamp)
+            throws Catalog.TableNotExistException {
+        Table table = catalog.getTable(Identifier.fromString(tableId));
+        FileStoreTable fileStoreTable = (FileStoreTable) table;
+        Snapshot snapshot = fileStoreTable.snapshotManager().earlierOrEqualTimeMills(timestamp);
+        Preconditions.checkNotNull(
+                snapshot, String.format("count not find snapshot earlier than %s", timestamp));
+        fileStoreTable.rollbackTo(snapshot.id());
+
+        return new String[] {"Success"};
```
Comment (LinMingQiang) [inline paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java:51]: Show which snapshot id to roll back to

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            })
+    public String[] call(ProcedureContext procedureContext, String tableId, Long timestamp)
+            throws Catalog.TableNotExistException {
+        Table table = catalog.getTable(Identifier.fromString(tableId));
+        FileStoreTable fileStoreTable = (FileStoreTable) table;
+        Snapshot snapshot = fileStoreTable.snapshotManager().earlierOrEqualTimeMills(timestamp);
+        Preconditions.checkNotNull(
+                snapshot, String.format("count not find snapshot earlier than %s", timestamp));
+        fileStoreTable.rollbackTo(snapshot.id());
+        return new String[] {"Success"};
```
Comment (LinMingQiang) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java:58]: ditto

File: docs/content/flink/procedures.md (hunk 1 above)
Comment (xuzifu666) [reply docs/content/flink/procedures.md:380]: My mistake, Thanks! @LinMingQiang

File: paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java (hunk 4 above)
Comment (xuzifu666) [reply paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java:51]: Good reminder~

File: paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ *  CALL sys.rollback_to_timestamp('tableId', timestamp)
+ * </code></pre>
+ */
+public class RollbackToTimestampProcedure extends ProcedureBase {
+
+    public static final String IDENTIFIER = "rollback_to_timestamp";
+
+    public String[] call(ProcedureContext procedureContext, String tableId, long timestamp)
+            throws Catalog.TableNotExistException {
+        Table table = catalog.getTable(Identifier.fromString(tableId));
```
Comment (LinMingQiang) [inline paimon-flink/paimon-flink-1.18/src/main/java/org/apache/paimon/flink/procedure/RollbackToTimestampProcedure.java:44]: check table not empty

**Comment by xuzifu666** [comment]
@LinMingQiang had addressed, Thanks

**Comment by LinMingQiang** [comment]
+1

**Review by JingsongLi — APPROVED** [review]
Thanks @xuzifu666 and @LinMingQiang

**Pull request merged**